### DISCUSSION TITLE
Minio interference with file system caching?

### DISCUSSION THREAD

**Discussion by aaime** [minio/minio#17686] [Q&A] [answered]
Hi all,
I'm comparing read performance through Minio vs reading directly from file system and found something strange.
Of course, HTTP access through Minio cannot be faster than the file system, but:

- If I read the file directly, eventually the operating system caches the (1.5GB) file in memory, and no disk access occurs any more
- If I read the same file through Minio, the disk is read every single time.

I find the Minio behavior surprisping... Minio is reading the file too, so the OS should be caching the reads for it as well, and instead, it's not happening. This is going to slow down repeated reads on a "hot" file that many clients want.

Is there anything that can be done, to convince the operating system to cache the reads when they are dong through Minio?
Btw, I'm currently doing local testing on Linux Mint 21.1 and I'm using the S3 interface for direct access to the file (it's not a signed link, it's direct HTTP access, S3 style, for a public bucket)

**Comment by harshavardhana** [comment, marked as answer]
We do not rely on file system caching, we always read directly from the storage device via O_DIRECT.

If you want you can turn off this feature and rely on filesystem cache if you want. Remember at scale filesystem cache is not very useful and instead it is slow.

```
mc admin config set alias api disable_odirect=on
```

**Reply by aaime** [reply]
Interesting, good to know.

The idea that the file system cache always slows down system sounds odd, at least with my GIS/cartography background. 
I can understand it being a bottleneck when accesses are randomly distributed to the point that no caching actually takes places, but in cartography/weather forecasting/satellite imagery that's hardly the case:
* Organizations keep a long time series of data for later extraction, but traffic concentrates on fresh data (which can be stored in Minio as a binary file, one per timestamp)
* If the data is split (also) across space, areas where there is human activities (ports, large cities) are typically the focus of attention, while deserts/oceans/polar areas are rarely visited. The same can be said for natural disasters, which tend to concentrate most access to the affected area.

Long story short, in my line of work there is a tendency to have lots of data, but few data sets account for the majority of traffic, and systems are given enough memory to leverage that locality of access, with a very significant speed-up.

**Reply by harshavardhana** [reply]
Filesystem caches quickly run out at scale. Think of scale-like 10 billion objects. Individual node filesystem cache has no way to store them in the cache. Filling the cache is an expensive operation than directly served from the disk.

If you repeatedly access the same object then you can turn off O_DIRECT but in reality, for most common workloads it is a bad idea.

**Reply by harshavardhana** [reply]
If you want to discuss this further how it affects production use-cases we are happy to get on a call please write to us hello@min.io

**Reply by julienlau** [reply]
Hi,
do you mean expensive in terms of CPU ?
When using an HDD setup and suffering for severe lack of IOPS with mean CPUwait ~15%, I don't think that is a valid argument. 
I have the feeling that this odirect option should always be disabled as long as we are on Read heavy workloads without NVMe.

**Reply by c4m4** [reply]
> We do not rely on file system caching, we always read directly from the storage device via O_DIRECT.
> 
> If you want you can turn off this feature and rely on filesystem cache if you want. Remember at scale filesystem cache is not very useful and instead it is slow.
> 
> ```
> mc admin config set alias api disable_odirect=on
> ```

Using version `20240507064125.0.0-1` with default odirect=on, I don't see  open system call using O_DIRECT flag.
